### PR TITLE
Enable cell editing and updating the table state

### PR DESCRIPTION
**Pull request by emirror-de**
This PR introduces the possiblity to edit cells and update the tables state automatically.

# Changes

1. **(Breaking)** Custom renderers now have to add a generic and two parameters to their components, `on_change: C` (`C` as an example of the new generic) and `editable: bool`.
1. As suggested in #3 a new method `fn set_row(&mut self, index: usize, row: T)` has been added to `TableDataProvider` trait.
1. The `date_cell_renderer` macro has been extended by a `default_date_format` parameter that is used to get the date format to be parsed for the specific date type when no custom format is available.
1. A new `editable` property of the `table` attribute has been added and can be applied to a struct member. If done, the table cell will get the `contenteditable` HTML property and update the tables state on the HTML `oninput` event.

# Discussion

The introduced implementation lacks the `async` possibility of the `set_row` method. Until now I do not have a good idea how to transfer the following line to async code:
```rust
row_data_provider.update_untracked(|d| d.set_row(i, new_item));
```
I tried to use an action for updating the row, but because the data provider is behind a `RwSignal` I was not able to get a reference and move it to an async block. Basically the problem is the following line that would be written in the action:
```rust
row_data_provider.update_untracked(|d| d.set_row(i, new_item).await);
```
This is of course not possible. Therefore this implementation needs to be adjusted to somehow add the possibility of setting the row using an async method. If there is no straight forward way to enable this, I am happy to mark this PR as complete from my point of view.

Looking forward to your ideas and comments.

### CODE REVIEW

**Review by maccesch — COMMENTED** [review]
Thanks for your great work! I found a few nitpicks but I like the overall solution.

**Comment by maccesch** [comment]
I'd like to think a bit about the async setter because for remote data sources this would be ideal.

File: src/components/cell/chrono.rs [hunk 1]
```diff
@@ -26,12 +28,40 @@ macro_rules! date_cell_renderer {
...
+                                None => $date_type::parse_from_str(&value, $default_date_format),
+                            };
+                            if let Ok(v) = parse_result {
+                                on_change(v);
+                            }
+                        }
+                    };
+
+                    return view! { cx,
+                        <td class=class _ref=td_ref on:input=on_input contenteditable>{text}</td>
```
[the remark this answers is not in the document]
Comment (emirror-de) [reply src/components/cell/chrono.rs]: Thanks for the hint, I have not been aware of this. I updated the PR accordingly,  [6a34bb8](https://github.com/Synphonyte/leptos-struct-table/pull/4/commits/6a34bb88ae718ef664e5fd1df6f922bb8da9d867).

**Comment by emirror-de** [comment]
> I'd like to think a bit about the async setter because for remote data sources this would be ideal.

This would of course be best to get this done. I have been running several times in a situation when it would be required to use a signal `with` or `update` method and an async method in it. The only way I currently see is to store the `DataProvider` in an `Arc<Mutex<_>>>`  but this would require changing a bit more in the code base. What do you think?

**Comment by emirror-de** [comment]
> Yeah I thought about sth similar. I'm not sure but do you think a stored value could help? https://docs.rs/leptos/latest/leptos/fn.store_value.html

Hmm there is an issue (https://github.com/leptos-rs/leptos/issues/1018) which is similar. They solve it with a `StoredValue` but it seems to me that it is not applicable for this case. :thinking:

**Comment by emirror-de** [comment]
I just gave it a try with `Arc<Mutex<_>>` but that does not make sense, because the `enum_items` resource does not work anymore. This is due to the mutex guard does not implement `PartialEq`. So another approach is required.

**Comment by maccesch** [comment]
Do we really need an `Arc<Mutex>`? Wouldn’t a `Rc<RefCell>` suffice? Again might be stupid I’m pretty new to Rust async.

**Comment by emirror-de** [comment]
So here is a big update and breaking change.

1. The `trait` has been split up into two, one for the provider, the other for the data. The user now is required to implement the provider by himself. With this approach it is possible to also provide implementations that have a connection to some remote backend.
2. A default implementation `MemoryStorage<T>` is available that provides a generic implementation for every type with a storage in memory.
3. Fields marked as `key` cannot be `editable` at the same time
4. The derived component contains a local state of the data that has been fetched. When edited, the local state is changed and the provider is notified to update.
5. The data is only fetched again when the `range` changes, as this is the only reason why the provider data should diverge from the local state. Sorting is considered to be only a visual change, not a change in the data. Therefore this has been removed from the dependencies for fetching the data.

I somehow broke the `getter` example, but I was not able to find the reason yet. It must happen somewhere [here](https://github.com/Synphonyte/leptos-struct-table-macro/pull/1/files#diff-b3ededbff3c9dee5c89919eebb93bc0989228772775166e566ab74325caf8ca3R258-R279).

**Comment by emirror-de** [comment]
Hm I just saw the example of the server function. I seem to have misunderstood the use of `set_sorting`. There might be an adjustment required with this.

**Comment by maccesch** [comment]
Thanks for all the work!
I kind of feel that this is a bit of a step backward from what I was trying to archieve with the data provider. I totally understand though why this wasn't clear to begin with. Let me try to explain...

The idea was to have a data interface that abstracts away where the data is coming from. Memory (`Vec`), Database, API, ...

To that end we don't always know all of the data but maybe only a certain slice of it. In the case of data coming from outside (e.g. Database, API) we only load the data necessary. That means sorting is being off-loaded to these external data sources, too, since we no longer can do it without knowing all of the data. That makes sorting intertwined with data fetching unfortunately.

**Comment by emirror-de** [comment]
Thanks for the explanations! So if I did get you right, the `set_sorting` method needs to move back into the `DataProvider` trait, because it already set the sorting on the external provider site?

As far as I know then the only thing I need to change again is:
- Rename `TableDataStorage` to `TableDataProvider`
- Integrate `set_sorting` to `TableDataProvider`

Shouldn't this automatically solve the async problem as well?

**Comment by maccesch** [comment]
Yes to all questions, I think 😄 

I don't know if that solves the async problem because with the changes you mention we're back at the same kind of `TableDataProvider` as before, right?

Thanks!

**Comment by emirror-de** [comment]
I just had a look at it again and I obviously forgot why I was splitting the trait into two.

The problem is that the `set_sorting` method requires the associated type `ColumnName`. This is generated by the macro crate.
My intention was to move the generic implementation for `TableDataProvider` to `leptos-struct-table` crate to increase readability and minimize the code generated by macro. Because the `ColumnName` enum is not available without the macro crate, I split the trait into two. So in contrary to what I wrote in the previous comment, recombining the traits is not possible.

There might be a "workaround" for it by making `TableDataSorting` a trait bound of `TableDataStorage` which is then implemented by the macro crate. I will test that locally.

Maybe this could be a compromise? What do you think?

**Comment by emirror-de** [comment]
> To that end we don't always know all of the data but maybe only a certain slice of it. In the case of data coming from outside (e.g. Database, API) we only load the data necessary. That means sorting is being off-loaded to these external data sources, too, since we no longer can do it without knowing all of the data. That makes sorting intertwined with data fetching unfortunately.

I do get your point. So maybe we should differentiate between sorting `locally` on the clients end and `off-loaded` on the data providers end.

Do I get you right, that when you change sorting the data is being fetched again from the data provider? Even if the sorting is only required locally?

If yes, we might be required to split these cases because this could lead to changing results in the table when the user does not expect them in my opinion. For example we show an excerpt of rows 10 to 20 without sorting. The user clicks on the first column. What the user would expect to see would be to just get these 10 rows sorted by the first column. Instead, the dataset is sorted on the providers end by the first column and after sorting the rows 10 to 20 are shown.
In addition to that it increases (for this case) the communication between client and provider without necessity.

**Comment by maccesch** [comment]
I'm closing this in favor of 34db429. 

Thanks for all the work! 
After a lot of thinking about this I decided it would be most idiomatic to provide an event handler to users if they want to change data.

Please see the editable example. I'm updating the docs right now as well.

The data provider and sorting logic are intimately related and thus should stay in the same trait IMO. If the user doesn't need sorting he is not required to implement that method.

**Pull request closed without merging**